### PR TITLE
Fix HTTP PeerCred authentication for domain users (fixes #1001)

### LINKED ISSUE #1001: CUPS in domain
We are trying to use CUPS on Linux machine connected to the MS AD domain.

When user logs into the system, it receives `uid=1136733317` or similar and this UID is properly resolved by the `sss` plugin to `/etc/nsswitch.conf` as `testovmt@example.com` with all appropriate groups membership, including the `lpadmin` group.

However, when user attempts to do some administrative tasks with printing system, the attempt is rejected as unauthorized and we see the following lines in log.

```
[04/Jul/2024:09:55:03 +0300] [Client 9] con->uri="/admin/", con->best=0x5580a2c09430(/admin)
D [04/Jul/2024:09:55:03 +0300] [Client 9] Authorized as testovmt@example.com using PeerCred.
d [04/Jul/2024:09:55:03 +0300] cupsdIsAuthorized: con->uri="/admin/", con->best=0x5580a2c09430(/admin)
d [04/Jul/2024:09:55:03 +0300] cupsdIsAuthorized: level=CUPSD_AUTH_USER, type=Basic, satisfy=CUPSD_AUTH_SATISFY_ALL, num_names=1
d [04/Jul/2024:09:55:03 +0300] cupsdIsAuthorized: auth=CUPSD_AUTH_ALLOW...
D [04/Jul/2024:09:55:03 +0300] cupsdIsAuthorized: username="testovmt@example.com"
d [04/Jul/2024:09:55:03 +0300] cupsdIsAuthorized: Checking user membership...
d [04/Jul/2024:09:55:03 +0300] cupsdCheckGroup(username="testovmt", user=(nil), groupname="lpadmin")
d [04/Jul/2024:09:55:03 +0300] cupsdCheckGroup(username="testovmt", user=(nil), groupname="root")
d [04/Jul/2024:09:55:03 +0300] cupsdCheckGroup(username="testovmt", user=(nil), groupname="wheel")
d [04/Jul/2024:09:55:03 +0300] [Client 9] cupsdSendError code=403, auth_type=0
D [04/Jul/2024:09:55:03 +0300] [Client 9] cupsdSendHeader: code=403, type="text/html", auth_type=0
D [04/Jul/2024:09:55:03 +0300] [Client 9] Closing connection.
```

Looking to the `scheduler/auth.c` source file, at around line 2205, I see the following code fragment:

```
 /*
  * Strip any @domain or @KDC from the username and owner...
  */

  if ((ptr = strchr(username, '@')) != NULL)
    *ptr = '\0';
     
  if (owner)
  {   
    strlcpy(ownername, owner, sizeof(ownername));

    if ((ptr = strchr(ownername, '@')) != NULL)
      *ptr = '\0';                      
  }   
  else
    ownername[0] = '\0';

```
So after successfully verifying that the supplied information via `HTTP PeerCred` corresponds to the `SO_PEERCRED` `UID`, CUPS proceeds to strip the domain part of the user name. Subsequently, only the local part of the user name is checked for group membership, mirroring the behavior observed in the logs.

This process leads to two significant issues:

1. Domain users may not be granted permissions that they actually possess, which is currently being experienced.
2. In cases where the local part of a domain user's name coincides with a local user's name, there is a risk of unintended permission escalation, posing a security vulnerability.

I suspect that there is a valid reason behind the decision to strip the domain part of the user name before group membership checks. Despite my efforts to trace this logic through the git history, it appears to originate from Apple's old SVN, which is inaccessible to the public.

Therefore, the key question is whether we should eliminate the domain name stripping process. If not, what would be the appropriate logic to implement in its place?

P.S.

There is a similar issue at the Red Hat Customer portal: https://access.redhat.com/discussions/6678951

### PR DESCRIPTION
**Pull request by lazyDesertMan**
- Remove domain from user name during local user authentication (e.g., "user@example.com" –> "user"). This practice can be beneficial for maintaining compatibility with older versions of Kerberos. However, enabling this option can have negative consequences. It may result in confusion between domain and local users with identical names, potentially leading to incorrect assignment of user permissions and unintentional permission escalation, thus creating a security risk. Therefore, it is advisable to avoid using this option in most cases.
- Add "StripUserDomain" parameter to cupsd.conf

### CODE REVIEW

**Review by alexpevzner — DISMISSED** [review]
I have reviewed this change, it looks good for me.

It resolves #1001 by adding an option that control stripping domain part of user name for user authorization purposes.

@michaelrsweet, @tillkamppeter, what are the next steps to have this PR merged into the CUPS?

File: doc/help/man-cupsd.conf.html
Comment (michaelrsweet) [inline doc/help/man-cupsd.conf.html:423]: This needs to be in the cups-files.conf man page, not cupsd.conf... Also, you need to edit the man page source, not the HTML (which gets generated from the man page source...)

"man/cups-files.conf.5"

File: scheduler/conf.c
Comment (michaelrsweet) [inline scheduler/conf.c:3385]: Don't add manual code for boolean options. Just add the StripUserDomain directive to the `cupsfiles_vars` array near the beginning of this source file with the `CUPSD_VARTYPE_BOOLEAN` annotation, e.g.:

```c
  { "StripUserDomain",	&StripUserDomain,	CUPSD_VARTYPE_BOOLEAN },
```

File: scheduler/conf.h
Comment (michaelrsweet) [inline scheduler/conf.h:182]: Don't specify a line number or a source file in the comment - that will become easily out of date. Better to say "Strip domain in local username?" and leave it at that. The man page will explain things in more depth plus we have comments where this is used.

**Review by michaelrsweet — CHANGES_REQUESTED** [review]
See comments - mostly correct but you need to move some things around.

Thanks!

File: doc/help/man-cupsd.conf.html
Comment (alexpevzner) [reply doc/help/man-cupsd.conf.html:423]: Hi @michaelrsweet,

do we need to update `conf/cups-files.conf.in` as well?

Comment (michaelrsweet) [reply doc/help/man-cupsd.conf.html:423]: You could add it (commented out) to the cups-files.conf template, but it isn't required/necessary.

**Comment by alexpevzner** [comment]
@michaelrsweet,

I don't quite understand what's next. I see PR was updated, probably resolving all your comments, but 1 unresolved requested change is pending in the conversation. It is not clear for me, how to mark it as resolved.

Should we just re-request a review from you?

**Comment by DiPush01** [comment]
Updated the indentation in the **conf.c** file in the new array entry. Sorry, we didn't notice right away that it required one tab instead of two.

**Review by michaelrsweet — CHANGES_REQUESTED** [review]
A couple minor changes and this will be ready to merge. Thanks!

**Comment by lazyDesertMan** [comment]
@michaelrsweet

Sending an updated version of the patch.

A small question remains. Does the request to move StripUserDomain over SyncOnClose only concern the conf.cpp file or should it have been done in the conf.h file as well?

**Comment by michaelrsweet** [comment]
@lazyDesertMan 

> A small question remains. Does the request to move StripUserDomain over SyncOnClose only concern the conf.c file or should it have been done in the conf.h file as well?

Just in the conf.c file.

**Comment by alexpevzner** [comment]
@michaelrsweet,

looks like everything is resolved. Can we merge now?

**Review by michaelrsweet — APPROVED** [review]
Looks good now, thank you!

**Comment by alexpevzner** [comment]
@michaelrsweet,

thank a lot!

**Comment by DiPush01** [comment]
@michaelrsweet,

Thanks for the review, can we backport it to 2.4.x?

**Comment by michaelrsweet** [comment]
@DiPush01 I'll leave that up to @zdohnal who is the 2.4.x release manager.

Zdenek, what do you think of backporting this (relatively simple) change?

**Comment by alexpevzner** [comment]
Hi @zdohnal,

I would personally ask you to accept this change to 2.4.x. We are running thousands of Linux installation across all public schools of a big city. #1001 is a real problem for us. This patch fixes that issue, and we want to minimize a difference of our CUPS sources from the upstream.

I believe, the millions of RedHat users will also benefit from this change.

**Comment by zdohnal** [comment]
@alexpevzner @michaelrsweet - The change looks simple, compiles in 2.4.x and tests passes on my local machine. Web docs were not updated in the PR, but I can add it in master as well.
The only thing which can be problematic is the different behavior than before - someone might depend on the previous behavior (IIUC CUPS removed the username and owner automatically, now it does only once StripUserDomain is on), so it might not be suitable for patch update as 2.4.11 will be, but rather for minor update 2.5.

However, since IMHO the change improves security, I will append it into release 2.4.11 - but be aware the if you provide this to your users/customers, note in documentation might be handy to warn people depending on old behavior.

**Comment by alexpevzner** [comment]
@zdohnal,

thank a lot!

We have already deployed this patch among our users, so far, there is no problems at all. The old behaviour looks IMHO really broken, I don't see any meaningful circumstances when `user@example.com` should be treated as a local `user`. May be only for compatibility with the really old Kerberos which has gone away years ago.

Do we need to prepare a separate RP for 2.4.x, you will do it by yourself?

**Comment by zdohnal** [comment]
@alexpevzner I have already pushed the change into the branch, but I referenced the original issue you had filed - #1001 .

**Comment by alexpevzner** [comment]
@zdohnal 

thank you very much!

**Comment by crysman** [comment]
Unfortunately, this has not been backported to Ubuntu 24.04.4 LTS ( cups 2.4.7-1.2ubuntu7.14).
If you want to help, can do "affects me too" in the bug ticket https://bugs.launchpad.net/ubuntu/+source/cups/+bug/2165919/+affectsmetoo

**Pull request merged**
